be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala

```diff
@@ -42,66 +51,86 @@ object GetMessagesByIdResponseHandler extends AskPatternConstants {
     case value @ _ => value
   }
 
-  // This function will try to digest the set of messages for an entire channel
-  // It will go channel by channel until it finished processing messages or MAX_RETRY is reached
+  /** Will try to digest each GraphMessage until their retry-counter reaches 0 or they all get validated
+    *
+    * @param receivedResponse
+    *   The wrapped messages with their counter
+    * @param validatorFlow
+    *   The validator to push the messages through
+    * @param system
+    *   Implicit actor system to use the given validator
+    * @return
+    *   true if all messages could be validated, otherwise false
+    */
   @tailrec
   private def passThroughPipeline(
-      receivedResponse: Map[Channel, Set[GraphMessage]],
-      validatorFlow: Flow[GraphMessage, GraphMessage, NotUsed],
-      remainingAttempts: Int
+      receivedResponse: List[(GraphMessage, Int)],
+      validatorFlow: Flow[GraphMessage, GraphMessage, NotUsed]
   )(implicit system: ActorSystem): Boolean = {
-    if (receivedResponse.isEmpty || remainingAttempts <= 0) {
+    if (receivedResponse.isEmpty || receivedResponse.forall(elem => elem._2 <= 0)) {
       return receivedResponse.isEmpty
     }
 
-    var failedMessages: Map[Channel, Set[GraphMessage]] = Map.empty
+    var failedMessages: List[(GraphMessage, Int)] = Nil
     receivedResponse.foreach {
-      case (channel, messagesSet) =>
-        val failedMsgErrorMap = messagesThroughPipeline(validatorFlow, messagesSet)
-        if (failedMsgErrorMap.nonEmpty) {
-          failedMessages += channel -> failedMsgErrorMap.keySet
-
-          // only log a during last attempt
-          if (remainingAttempts == 1) {
-            println("Errors: " + failedMsgErrorMap.map {
-              case (msg, error) => error.toString + "\n On:\n" + prettyPrinter(msg)
-            })
+      case (graphMessage, retry) =>
+        if (retry > 0) {
+          messagesThroughPipeline(validatorFlow, graphMessage) match {
+            case Left(err) =>
+              failedMessages = (graphMessage, retry - 1) :: failedMessages
```

Comment:
Nitpicks: You could probably use the operator `::=` that you made me discover a couple of lines above :)
On a side note, maybe using a mutable collection from `scala.collection.mutable` would be better than using a `var` for performance purposes